### PR TITLE
Versatile WebSocket EDGECLOUD-3142

### PR DESCRIPTION
**Pull request by ah1053**
Versatile WebSocket:
1. Sending Binary and Text Data (2 threads and two queues).
2. Receiving Binary and Text Data ( 1 thread and two queues).
3. Name change, new name is (MobiledgeXWebSocketClient.cs).

Tested on ComputerVision

### CODE REVIEW

**Comment by franklin-huang-mobiledgex** [comment]
@lgarner Is there a situation where a Websocket client may want to send String msg and Binary msg to the same host and port?

**Comment by brucearmstrong** [comment]
> @lgarner Is there a situation where a Websocket client may want to send String msg and Binary msg to the same host and port?

The ComputerVision server can accept either binary or text data, both on the same host, port, and URL. For binary, we operate on it as an image. For text, we simply echo the received text back. This is used for a full round trip "ping" time.

File: Runtime/Scripts/MobiledgeXWebSocketClient.cs [hunk 1]
```diff
@@ -108,9 +89,14 @@ public void Send(string message)
...
             //Debug.Log("Message to queue for send: " + buffer.Length + ", message: " + message);
             var sendBuf = new ArraySegment<byte>(buffer);
-
             sendQueue.Add(sendBuf);
         }
+       public async void Send(byte[] binary)
+        {
+            //Debug.Log("Message to queue for send: " + buffer.Length + ", message: " + message);
+            var sendBuf = new ArraySegment<byte>(binary);
+            await ws.SendAsync(sendBuf, WebSocketMessageType.Binary, true /* is last part of message */, CancellationToken.None);
```
Comment (franklin-huang-mobiledgex) [inline Runtime/Scripts/MobiledgeXWebSocketClient.cs:98]: Is there no sendQueue for binary data?

Comment (ah1053) [reply Runtime/Scripts/MobiledgeXWebSocketClient.cs:98]: I wanted to use the MobiledgeXWebSocketClient in the CV app so small change to send, the sending queue is of type string.

File: Runtime/Scripts/MobiledgeXWebSocketClient.cs [hunk 2]
```diff
@@ -108,9 +89,14 @@ public void Send(string message)
             byte[] buffer = encoder.GetBytes(message);
             //Debug.Log("Message to queue for send: " + buffer.Length + ", message: " + message);
             var sendBuf = new ArraySegment<byte>(buffer);
-
             sendQueue.Add(sendBuf);
         }
+       public async void Send(byte[] binary)
```
Comment (lgarner) [inline Runtime/Scripts/MobiledgeXWebSocketClient.cs:94]: Line.

File: Runtime/Scripts/MobiledgeXWebSocketClient.cs (hunk 1 above)
Comment (lgarner) [reply Runtime/Scripts/MobiledgeXWebSocketClient.cs:98]: The websocket client has independent queues for send and receive, await would freeze the UI sending that binary buffer.

Comment (ah1053) [reply Runtime/Scripts/MobiledgeXWebSocketClient.cs]: @lgarner  @franklin-huang-mobiledgex  I added another queue for binary messages and an independent thread (sendThreadBinary) that is being initialized at the constructor.

**Comment by franklin-huang-mobiledgex** [comment]
This looks better. Should we add a parameter to the Receive function that indicates whether the developer wants binary data or string data? Developer may must want raw bytes instead of the String conversion

File: Runtime/Scripts/MobiledgeXWebSocketClient.cs [hunk 3]
```diff
@@ -149,31 +172,48 @@ public async Task<string> Receive(UInt64 maxSize = MAXREADSIZE)
...
             {
                 //Debug.Log("Awaiting Receive...");
-                result = await Receive();
-                if (result != null && result.Length > 0)
+                response = await Receive();
+                if (response != null && response.Keys.Count > 0)
                 {
                     //Debug.Log("Received: " + result);
-                    receiveQueue.Enqueue(result);
+                    if (response.Keys.First() == WebSocketMessageType.Text)
```
Comment (franklin-huang-mobiledgex) [inline Runtime/Scripts/MobiledgeXWebSocketClient.cs:199]: how does the developer set the receive type?

Comment (ah1053) [reply Runtime/Scripts/MobiledgeXWebSocketClient.cs:199]: The developer doesn't need to set the receive type, The receive type is dedicated by the server.

Comment (ah1053) [reply Runtime/Scripts/MobiledgeXWebSocketClient.cs:199]: as discussed
 chunkResult is of type WebSocketReceiveResult not byte[]
WebSocketReceiveResult chunkResult >> WebSocketReceiveResult has MessageType field

**Pull request merged**